receiver/googlecloudmonitoringreceiver/receiver.go

```diff
@@ -5,51 +5,229 @@ package googlecloudmonitoringreceiver // import "github.com/open-telemetry/opent
 
 import (
 	"context"
+	"errors"
+	"fmt"
+	"os"
+	"sync"
+	"time"
 
+	monitoring "cloud.google.com/go/monitoring/apiv3/v2"
+	"cloud.google.com/go/monitoring/apiv3/v2/monitoringpb"
 	"go.opentelemetry.io/collector/component"
 	"go.opentelemetry.io/collector/pdata/pmetric"
 	"go.uber.org/zap"
+	"google.golang.org/api/iterator"
+	"google.golang.org/api/option"
+	"google.golang.org/genproto/googleapis/api/metric"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/googlecloudmonitoringreceiver/internal"
 )
 
 type monitoringReceiver struct {
-	config *Config
-	logger *zap.Logger
-	cancel context.CancelFunc
+	config         *Config
+	logger         *zap.Logger
+	client         *monitoring.MetricClient
+	metricsBuilder *internal.MetricsBuilder
+	mutex          sync.Mutex
 }
 
 func newGoogleCloudMonitoringReceiver(cfg *Config, logger *zap.Logger) *monitoringReceiver {
 	return &monitoringReceiver{
-		config: cfg,
-		logger: logger,
+		config:         cfg,
+		logger:         logger,
+		metricsBuilder: internal.NewMetricsBuilder(logger),
 	}
 }
 
-func (m *monitoringReceiver) Scrape(ctx context.Context) (pmetric.Metrics, error) {
-	// Dummy use to fix lint errors
-	ctx.Deadline()
-	metrics := pmetric.NewMetrics()
-	m.logger.Debug("Scrape metrics ")
+func (mr *monitoringReceiver) Start(ctx context.Context, _ component.Host) error {
+	var client *monitoring.MetricClient
+	var err error
 
-	return metrics, nil
-}
+	// Lock to ensure thread-safe access to mr.client
+	mr.mutex.Lock()
+	defer mr.mutex.Unlock()
+
+	// If the client is already initialized, return nil
+	if mr.client != nil {
+		return nil
+	}
+
+	// Get service account key file path
+	serAccKey := os.Getenv("GOOGLE_APPLICATION_CREDENTIALS")
+	if serAccKey != "" {
+		// Use provided credentials file
+		credentialsFileClientOption := option.WithCredentialsFile(serAccKey)
+		client, err = monitoring.NewMetricClient(ctx, credentialsFileClientOption)
+	} else {
+		// Set default credentials file path for testing
+		os.Setenv("GOOGLE_APPLICATION_CREDENTIALS", "testdata/serviceAccount.json")
 
-func (m *monitoringReceiver) Start(ctx context.Context, _ component.Host) error {
-	ctx, m.cancel = context.WithCancel(ctx)
-	err := m.initialize(ctx)
+		// Fallback to Application Default Credentials(https://google.aip.dev/auth/4110)
+		client, err = monitoring.NewMetricClient(ctx)
+	}
+
+	// Attempt to create the monitoring client
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to create a monitoring client: %w", err)
 	}
+	mr.client = client
+	mr.logger.Info("Monitoring client successfully created.")
+
+	// Return nil after client creation
 	return nil
 }
 
-func (m *monitoringReceiver) Shutdown(context.Context) error {
-	m.logger.Debug("shutting down googlecloudmonitoringreceiver receiver")
-	return nil
+func (mr *monitoringReceiver) Shutdown(context.Context) error {
+	var err error
+	if mr.client != nil {
```

Comment:
I think you should probably lock around this.

Reply:
Updated